### PR TITLE
add stein methods

### PR DESCRIPTION
**Pull request by ferrine**
More and more papers use stein methods. Here I implement convenient interface to it

### CODE REVIEW

File: pymc3/variational/inference.py [hunk 1]
```diff
@@ -19,6 +19,7 @@
     'ADVI',
     'FullRankADVI',
     'SVGD',
+    'ASVGD',
```
Comment (twiecki) [inline pymc3/variational/inference.py:22]: what's ASVGD?

Comment (ferrine) [reply pymc3/variational/inference.py:22]: Amortized Stein Variational Gradient Descent 
Can be found [here](http://bayesiandeeplearning.org/papers/BDL_21.pdf). It is a novel black box algorithm for *wild inference* that doesn't require q(z)

File: pymc3/variational/__init__.py [hunk 2]
```diff
@@ -1,39 +1,19 @@
...
-    Empirical,
-    FullRank,
-    MeanField,
-    sample_approx
-)
+from .inference import *
+from . import inference
 
-from .stein import Stein
+from .approximations import *
```
Comment (twiecki) [inline pymc3/variational/__init__.py]: I think we removed all * imports in other parts of the code-base, shouldn't reintroduce it here. cc @ColCarroll

**Comment by ferrine** [comment]
That's ready for review

**Comment by twiecki** [comment]
I assigned @taku-y if he has time.

**Comment by ferrine** [comment]
I think I'll add a note about behavior of sampler in high dimensions. After I see no problem for merge 
CC @taku-y

**Comment by ferrine** [comment]
Important Note:
For BEST model with FullRank approx I found an exact solution with large number of particles(1000). With low number of them variance was underestimated. So I'm not sure if it really works as proposed in the paper(theoretically). CC @lewisKit

Blue is ASVGD on FullRank approx, green is FullRankADVI
image

**Comment by taku-y** [comment]
Here is the log of Travis on `TestSVGD.test_optimizer_minibatch_with_generator`:

```
>       np.testing.assert_allclose(np.std(trace['mu']), np.sqrt(1. / d), rtol=0.4)
E       AssertionError: 
E       Not equal to tolerance rtol=0.4, atol=0
E       
E       (mismatch 100.0%)
E        x: array(0.13835382035008825)
E        y: array(0.09476178269858855)
```

I'm not sure if this error rate is acceptable.

**Comment by ferrine** [comment]
I'm more concerned about poor ASVGD performance on simple problems. I'l increase number of iterations for these tests and lower relative tolerance

**Comment by ferrine** [comment]
@taku-y I've fixed tests, they look better now.  (passing locally)

**Comment by taku-y** [comment]
ASVGD results are comparable on the BEST example and the code looks good to me. I think its ready to merge.

**Comment by twiecki** [comment]
Awesome! Should also add an example with short description.

**Comment by ferrine** [comment]
Sure, coming soon in my blog

**Comment by ferrine** [comment]
I'll point on the problems that can be met using asvgd.

**Comment by springcoil** [comment]
Excellent work all

**Pull request merged**